### PR TITLE
mtl semaphore release

### PR DESCRIPTION
**Pull request by hana-alice**
Re: #https://github.com/cocos/cocos-engine/pull/12185

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/renderer/gfx-metal/MTLSemaphore.h [hunk 1]
```diff
@@ -32,38 +32,52 @@ namespace gfx {
...
+    };
     CCMTLSemaphore(const CCMTLSemaphore &) = delete;
     CCMTLSemaphore(CCMTLSemaphore &&) = delete;
     CCMTLSemaphore &operator=(const CCMTLSemaphore &) = delete;
     CCMTLSemaphore &operator=(CCMTLSemaphore &&) = delete;
 
-    void signal() const {
+    void signal() {
         dispatch_semaphore_signal(_semaphore);
+        --_restCount;
```
Comment (dumganhar) [inline native/cocos/renderer/gfx-metal/MTLSemaphore.h]: OperatIng `int32_t _restCount;` in different thread is dangerous since it isn't atomic.

**Comment by fish9167** [comment]
@cocos-robot run test cases

File: native/cocos/renderer/gfx-metal/MTLSemaphore.h [hunk 2]
```diff
@@ -32,38 +32,52 @@ namespace gfx {
...
 public:
-    explicit CCMTLSemaphore(uint32_t initialValue) : _semaphoreCount(initialValue) {
+    explicit CCMTLSemaphore(uint32_t initialValue) : _semaphoreCount(initialValue), _restCount(initialValue) {
         _semaphore = dispatch_semaphore_create(initialValue);
     }
-    ~CCMTLSemaphore() = default;
+    
+    ~CCMTLSemaphore() {
+        if(_semaphore) {
+            while(_restCount-- >0) {
```
Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLSemaphore.h]: `_restCount ` is minus in `signal()`, duplicate?

Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLSemaphore.h]: `_restCount-- >0` -> `_restCount-- > 0`

File: native/cocos/renderer/gfx-metal/MTLSemaphore.h [hunk 3]
```diff
@@ -32,38 +32,52 @@ namespace gfx {
...
-    void wait() const {
+    void wait() {
         dispatch_semaphore_wait(_semaphore, DISPATCH_TIME_FOREVER);
+        ++_restCount;
     }
 
     void trySyncAll(uint64_t nanoSec) {
         for (uint32_t i = 0; i < _semaphoreCount; i++) {
             dispatch_semaphore_wait(_semaphore, nanoSec);
+            ++_restCount;
```
Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLSemaphore.h]: I think can just invoke `wait()`

File: native/cocos/renderer/gfx-metal/MTLSemaphore.h [hunk 4]
```diff
@@ -32,38 +32,52 @@ namespace gfx {
...
+    explicit CCMTLSemaphore(uint32_t initialValue) : _semaphoreCount(initialValue), _restCount(initialValue) {
         _semaphore = dispatch_semaphore_create(initialValue);
     }
-    ~CCMTLSemaphore() = default;
+    
+    ~CCMTLSemaphore() {
+        if(_semaphore) {
+            while(_restCount-- >0) {
+                dispatch_semaphore_signal(_semaphore);
+            }
```
Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLSemaphore.h:43]: May be can just do it like
```c++
while (_restCount) {
    signal();
}
```

File: native/cocos/renderer/gfx-metal/MTLSemaphore.h [hunk 5]
```diff
@@ -32,38 +32,52 @@ namespace gfx {
...
         for (uint32_t i = 0; i < _semaphoreCount; i++) {
             dispatch_semaphore_wait(_semaphore, nanoSec);
+            ++_restCount;
         }
     }
 
     void syncAll() {
         for (uint32_t i = 0; i < _semaphoreCount; i++) {
             dispatch_semaphore_wait(_semaphore, DISPATCH_TIME_FOREVER);
+            ++_restCount;
```
Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLSemaphore.h]: The same as above.

File: native/cocos/renderer/gfx-metal/MTLSemaphore.h [hunk 6]
```diff
@@ -32,38 +32,52 @@ namespace gfx {
 
 class CCMTLSemaphore final {
 public:
-    explicit CCMTLSemaphore(uint32_t initialValue) : _semaphoreCount(initialValue) {
+    explicit CCMTLSemaphore(uint32_t initialValue) : _semaphoreCount(initialValue), _restCount(initialValue) {
```
Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLSemaphore.h]: I think initial value of `_restCount ` is 0.

File: native/cocos/renderer/gfx-metal/MTLSemaphore.h [hunk 7]
```diff
@@ -32,38 +32,52 @@ namespace gfx {
...
         for (uint32_t i = 0; i < _semaphoreCount; i++) {
             dispatch_semaphore_wait(_semaphore, DISPATCH_TIME_FOREVER);
+            ++_restCount;
         }
     }
 
 protected:
     dispatch_semaphore_t _semaphore = nullptr;
     uint32_t _semaphoreCount = 0;
+    int32_t _restCount;
```
Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLSemaphore.h]: default value.

File: native/cocos/renderer/gfx-metal/MTLSemaphore.h (hunk 2 above)
Comment (hana-alice) [reply native/cocos/renderer/gfx-metal/MTLSemaphore.h]: https://man.archlinux.org/man/dispatch_semaphore_create.3.en to ensure signal at least 3 semphores when released.

**Review by minggo — APPROVED** [review]

**Review by dumganhar — APPROVED** [review]

**Pull request merged**